Add Nucleo_F746ZG support

Bob-the-Kuhn opened the pull request:
This PR adds support for the Nucleo_F746ZG development board.

The main value of this PR is it adds SDIO support for the STM32F7xx series.

No changes were needed for SPI access to the SD card.

Just for fun I added pin assignments for the REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER and similar devices.  A custom cable is needed to provide the needed +5V.

Testing hardware:
- Nucleo_F746ZG
- [Adafruit Micro SD SPI or SDIO Card Breakout Board](https://www.adafruit.com/product/4682)

Environments were added which support bootloader and non-bootloader systems.  

The bootloader environment needs some additional work.  Right now two variables need to be updated in order to change the build/load address. VECT_TAB_OFFSET is used by a low level init routine to set the VTOR.  I couldn't figure out how to use that variable to modify the load/build address so I created a LD file that has the offset hard coded into it.  The Marlin defined boards use board_build.offset or board_build.address to accomplish this.  Maybe the solution is to create a Marlin STMF32F7xx variant.

The new pins file has a BOARD_PREINIT macro in it.  It is used to init the ITCM ram or else The SDIO callbacks don't register correctly.

thisiskeithb commented:
> Just for fun I added pin assignments for the REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER and similar devices. A custom cable is needed to provide the needed +5V.

While I don't expect this board to be popular / actually used for a 3D printer, it'd be good to add an `#ifndef NO_CONTROLLER_CUSTOM_WIRING_WARNING` check & error that we've added to other pins files that require a custom LCD cable.

Bob-the-Kuhn commented:
Created a local/Marlin variant.  That solved the "unable to change load point" problem.

Added M43 support.  M42 and M43 need some more work.

Bob-the-Kuhn commented:
Looks like M43 is working correctly but M42 will assign the same PWM channel to multiple pins.  The result is multiple pins are locked together.  Changing one pin's PWM setting results in the other pin(s) behaving the same.

Bob-the-Kuhn commented:
Added a routine that keeps track of the PWM assignments and doesn't allow one PWM channel to be assigned to two different physical pins.

I'm not sure that's the best long term approach.  

Probably should add a debug print for the case when an attempt to assign one PWM channel to multiple pins is prevented.

Right now it's only enabled for the STM32F7xx CPU.  Need to consider if it should be dropped, more CPUs enabled, ...

---

Still need to back out my config file changes.  So far I'm 0 for 2 on fixing it.

Bob-the-Kuhn commented:
Finally was able to back out the inadvertent config file changes.

Added a wrapper to PWM_PINS to keep it from getting compiled for non-target boards.  This cleared up a Klemtest issue.

Bob-the-Kuhn commented:
Just did a little cleanup and minor changes:
1) consolidate #if logic in fastio.h
2) allow use of PWM_PIN with G0 and H7 parts.
3) add #if statements to PWM_PIN so that it could be used with smaller CPUs in the future.

I believe this PR is ready for review.

Still not sure if the PWM_PIN change should remain.  It's nice to have when setting up a new board but of no use to the average user.

thinkyhead commented on Marlin/src/HAL/STM32/pinsDebug.h:
```diff
@@ -280,9 +296,36 @@ void pwm_details(const pin_t Ard_num) {
+    if (WITHIN(Ard_num, 83, 90))
+      return Ard_num - 67;
+    return -1;
+  }
```
Double-checking the datasheet becuse this isn't exactly matching up…
`pinsDebug.h`:
```cpp
if (WITHIN(Ard_num, 0, 7))    return Ard_num;       // Pins  0 -  7  =>  A0  - A7
if (WITHIN(Ard_num, 16, 17))  return Ard_num - 8;   // Pins 16 - 17  =>  A8  - A9
if (WITHIN(Ard_num, 32, 37))  return Ard_num - 22;  // Pins 32 - 37  =>  A10 - A15
if (WITHIN(Ard_num, 83, 90))  return Ard_num - 67;  // Pins 80 - 90  =>  A16 - A23
```
`HAL.h`:
```cpp
if (WITHIN(Ard_num, 0, 7))    return Ard_num;      // A0  - A7   =>  Pins  0 -  7
if (WITHIN(Ard_num, 8, 9))    return Ard_num + 8;  // A8  - A9   =>  Pins 16 - 17
if (WITHIN(Ard_num, 11, 12))  return Ard_num + 22; // A11 - A12  =>  Pins 33 - 34
if (WITHIN(Ard_num, 13, 23))  return Ard_num + 67; // A13 - A23  =>  Pins 80 - 90
```
It looks like `pinsDebug.h` is probably the correct one since `HAL.h` is missing A10 altogether.

thinkyhead replied:
Correction pushed.

Bob-the-Kuhn replied:
I'm thinking a different approach is needed.    Trying to produce a simple **digital_pin_to_analog_ch**  function for each variant will be like herding cats.  

I'm going to look at scanning the  **digitalPin** and **analogInputPin** arrays to see if the requested pin is available to be assigned as an AIN.

Worst comes to worst just looking at the MODER for the requested pin and returning true or false rather than a channel number may produce all that a M43 user is usually interested in.

thinkyhead commented:
> Still not sure if the PWM_PIN change should remain. It's nice to have when setting up a new board but of no use to the average user.

We can retain it as a switchable option, as it seems like it could be useful for future development.

Bob-the-Kuhn commented:
Changed the approach used in the **digital_pin_to_analog_ch** function.

Looks like there are two methods used to map analog inputs to digital pin names used in the STM32 variants.  
- defines NUM_ANALOG_FIRST & NUM_ANALOG_INPUTS along with the array **digitalPin[]** map a contiguous range of pins to analog inputs.
- define NUM_ANALOG_INPUTS along with the arrays **analogInputPin[]** and **digitalPin[]** map multiple ranges of pins toanalog inputs.

Rather than trying to mimic variant specific implementations, switched to scanning the **analogInputPin** array.  The scanning method should be applicable to most of the variants.

Bob-the-Kuhn commented:
Still trouble shooting a bug on the STM32F746 code.  

The STM32F746 code is only allowing a subset of the available analog inputs to be mapped/used.  I was expecting to be able to assign any of the analog inputs as the TEMP_0_PIN.  Right now I can assign any input from the subset as the TEMP_0_PIN.

thinkyhead commented:
Is there something that you wanted to add in relation to the new environment for STM32H753ZI? It might be better to hold those changes until there's a board to go with it.

Bob-the-Kuhn commented:
I'm OK putting this PR on hold until a STM32H753ZI 3D printer controller is available,

thinkyhead commented:
I went ahead and added a test for this new environment, and it fails due to some of the system headers. Not sure how to proceed.

thinkyhead committed:
Add Nucleo_F746ZG support

Co-Authored-By: Bob Kuhn <20692583+Bob-the-Kuhn@users.noreply.github.com>
